packages/upload-client/test/index.test.js

```diff
@@ -1,8 +1,182 @@
 import assert from 'assert'
-import Client from '../src/index.js'
+import * as Client from '@ucanto/client'
+import * as Server from '@ucanto/server'
+import * as CAR from '@ucanto/transport/car'
+import * as CBOR from '@ucanto/transport/cbor'
+import * as Signer from '@ucanto/principal/ed25519'
+import * as StoreCapabilities from '@web3-storage/access/capabilities/store'
+import * as UploadCapabilities from '@web3-storage/access/capabilities/upload'
+import { uploadFile, uploadDirectory } from '../src/index.js'
+import { serviceSigner } from './fixtures.js'
+import { randomBytes } from './helpers/random.js'
+import { File } from './helpers/shims.js'
+import { mockService } from './helpers/mocks.js'
 
-describe('index', function () {
-  it('should export a client object', () => {
-    assert(Client)
+describe('uploadFile', () => {
+  it('uploads a file to the service', async () => {
+    const res = {
+      status: 'upload',
+      headers: { 'x-test': 'true' },
+      url: 'http://localhost:9200',
+    }
+
+    const account = await Signer.generate()
+    const issuer = await Signer.generate() // The "user" that will ask the service to accept the upload
+    const file = new Blob([await randomBytes(128)])
+    /** @type {import('../src/types').CARLink|undefined} */
+    let carCID
+
+    const proofs = await Promise.all([
+      StoreCapabilities.add.delegate({
+        issuer: account,
+        audience: serviceSigner,
+        with: account.did(),
+        expiration: Infinity,
+      }),
+      UploadCapabilities.add.delegate({
+        issuer: account,
+        audience: serviceSigner,
+        with: account.did(),
+        expiration: Infinity,
+      }),
+    ])
```

Comment:
How does this work? what ties issuer + account together? why is `issuer` allowed to invoke an upload with this proof chain?

Reply:
How does this work? what ties issuer + account together?

I'm not sure I understand what you mean, here also what test does here is hopefully not what we do in the actual client. Idea is that `space` should delegate all of it's capabilities, that is `{ can: *, with: space.did() }` to the service DID so that service in the future could delegate it back to the user.

Problem here is that it delegates just specific capabilities not all of them. That would imply that if user looses key we can delegate back only these two capabilities but nothing else.

> why is `issuer` allowed to invoke an upload with this proof chain?

This is precisely what the [ownership section](https://hackmd.io/Zb7gjpLsQn2w3a3JUvnFcw?view#Ownership) here tries to address.

If `with` DID is the same as `iss` DID you don't need any proof because it's self-evident that you own that resource, otherwise how would you would not be able to issue such UCAN because you would not be able to sign without a private key.

P.S.: If you think there is a good / better place way to capture that let me know

Reply:
If you mean how does space tied to the agent then

1. On space creation:
   a. space delegates `{ can: *, with: space.did() }` to `agent.did()` - So agent can act on user behalf
   b. space delegates `{ can: *, with: space.did() }` to `service.did()` - So service can delegate to users other agent after email verification. Key recovery & rotation is the same flow.
   c. space should delegate to user email as per #165 
2. On login to existing space
   a. User types email address and agent asks service for the delegation to (local) `agent.did()`
   b. Service delegates capabilities from 1. .b to requesting `agent.did()` and sends an email with a link to import it. After clicking it agent imports delegation and now has access to the space.

As an aside we would be able to replace delegation form space -> service, with space to user email address so we never actually have user space capabilities, because with [ucan-mailto](https://github.com/ucan-wg/ucan-mailto/pull/1/files) will be able to do email to did delegations

Reply:
my question was really "why is the issuer allowed to invoke `store/add` on this space when the proof chain the issuer provides doesn't include a delegation from space to issuer".

Reply:
this test passes and i assumed it would fail as it is currently written.

Reply:
> my question was really "why is the issuer allowed to invoke `store/add` on this space when the proof chain the issuer provides doesn't include a delegation from space to issuer".

Got it, does [this comment](https://github.com/web3-storage/w3protocol/pull/148#discussion_r1024296793) answers it / makes sense ?

Reply:
Very helpful, thanks! So if I understand correctly the self-evident proof is a very sensitive bunch o bytes. If it got intercepted any agent could use it... Any agent could upload to thay space if they got hold of such a proof. (Were it to be constructed in the wild)

Reply:
```js
StoreCapabilities.add.delegate({
  issuer: account,
  audience: serviceSigner,
  with: account.did(),
  expiration: Infinity,
}),
```

As I understood it when I wrote it, the space (`account` here) is delegating the `store/add` capability. The delegation is valid for use with our service (the audience) and with the passed space (the resource, or "with").

Anyone who recieves this delegation can issue `store/add` invocations for the space, at our service using this delegation as the proof.

So there's 2 questions:

1. Is that the right interpretation of what is being delegated and what is allows?
2. I changed `audience` in this delegation from the service to the agent and were able to successfully call the service with this delegation as proof. Why does that work? I expected ucanto server to reject. Is it a bug?

Reply:
> As I understood it when I wrote it, the space (`account` here) is delegating the `store/add` capability. The delegation is valid for use with our service (the audience) and with the passed space (the resource, or "with").
>
> Anyone who recieves this delegation can issue store/add invocations for the space, at our service using this delegation as the proof.
🎯

> Is that the right interpretation of what is being delegated and what is allows?

Yes

> I changed audience in this delegation from the service to the agent and were able to successfully call the service with this delegation as proof. Why does that work? I expected ucanto server to reject. Is it a bug?

It is correct because when you invoked capability you end up with following delegation chain

```
account -> agent -> service
```

account can self issue because it own's DID, agent can invoke because it has delegation from the account

Reply:
> Very helpful, thanks! So if I understand correctly the self-evident proof is a very sensitive bunch o bytes. If it got intercepted any agent could use it... Any agent could upload to thay space if they got hold of such a proof. (Were it to be constructed in the wild)

No it's not, self-evident delegation from owner is always to some DID and signed by the issuer (which is owner). No one other than DID owner delegated to could invoke that capability, because they would not be able to sign with the private key of that DID. If some other DID uses that delegation it will no longer be a valid chain it will be mole like

```
account -> agent -x- attacker -> service
```

Reply:
i'm still not getting it! In this test we do this:

```js
     const account = await Signer.generate()
     const issuer = await Signer.generate() // The "user" that will ask the service to accept the upload

     const proofs = await Promise.all([
       StoreCapabilities.add.delegate({
         issuer: account,
         audience: serviceSigner,
         with: account.did(),
         expiration: Infinity,
       }),
       UploadCapabilities.add.delegate({
         issuer: account,
         audience: serviceSigner,
         with: account.did(),
         expiration: Infinity,
       }),
     ])

    // ... trim server and conn creation ...

    const dataCID = await uploadFile({ issuer, proofs }, file, { connection } )
```

to my eyes that looks like we set up a proof that says 
- _"dear service, account here. allow store/add, upload/add on account, signed off by yours truly, account"_  ...with no mention of `issuer` the signer created on line 2 in the above snippet.

then it's invoked like
- _"hey service! issuer here! i'd like to store/add this car plz! here is a proof that says that account is cool with folks doing store/add and upload/add on account, so i guess store it in account. They didn't mention me explicitly, but i found this string of bytes accidentally sent to a log file, so i pasted it in here, but we're cool, right? it's all signed off by account! Also, I, issuer, am also obviously me (not account), as i have signed this off as `issuer`."_

Reply:
that maybe just how it works, i'm just trying to clarify my concern around such a self evident proof looking like maybe it is dangerously powerful. Why is a proof chain that never addresses our signer called `issuer` as the `audience` for the capbailities usable by `issuer` to prove they can do a thing?

Reply:
> i'm still not getting it! In this test we do this:
> 
> ```js
>      const account = await Signer.generate()
>      const issuer = await Signer.generate() // The "user" that will ask the service to accept the upload
> 
>      const proofs = await Promise.all([
>        StoreCapabilities.add.delegate({
>          issuer: account,
>          audience: serviceSigner,
>          with: account.did(),
>          expiration: Infinity,
>        }),
>        UploadCapabilities.add.delegate({
>          issuer: account,
>          audience: serviceSigner,
>          with: account.did(),
>          expiration: Infinity,
>        }),
>      ])
> 
>     // ... trim server and conn creation ...
> 
>     const dataCID = await uploadFile({ issuer, proofs }, file, { connection } )
> ```
> 
> to my eyes that looks like we set up a proof that says

I don't know what `uploadFile` does there but assuming it issues some invocation using `proofs` as proofs it should not work. Issuer there should be an `account` or delegations should change to delegate to issuer instead of `serviceSigner`

Reply:
Again it kind of depends what `uploadFile` does it may be using `issuer.did()` in `with` field in which case proofs are pointless and validator won't look at them because `issuer` could issue any capability as long as `with` is `issuer.did()` without proofs

Reply:
> that maybe just how it works, i'm just trying to clarify my concern around such a self evident proof looking like maybe it is dangerously powerful. Why is a proof chain that never addresses our signer called issuer as the audience for the capbailities usable by issuer to prove they can do a thing?

No if what happens under the hood looks roughly like the following, validator would have to say you're not authorized and if it does not I'm very surprised that such bug has escaped:


```ts
{ 
   iss: 'did:key:iss'
   aud: 'did:key:service'
   att: [
       { can: 'store/add', with: 'did:key:acc', link: 'bag...car' },
   ],
  proofs: [
      { iss: 'did:key:acc'
        aud: 'did:key:service'
        att: [{ can: "store/add", with: "did:key:acc" }]
      }
  ]
}
```

Reply:
I see now that what @olizilla was describing is actually what's in this test, I'm surprised it's passing. I have created test in ucanto to check for principal aligment https://github.com/web3-storage/ucanto/pull/134

I'll continue investigating why this isn't failing